### PR TITLE
James 1874 Solve GetMessageList slowliness

### PR DESCRIPTION
**Pull request by chibenwa**
A little improvement from Antoine,

Then not carring about this "recent" madness,

And finally *4 speedup for IMAP select on large mailboxes (sorry, the full list of recent is needed :-p)

I also worked to improve a bit code quality with an enhacement of Updated flags and adding a missing class of tests.

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentsDAO.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    private PreparedStatement createAddStatement(Session session) {
+        return session.prepare(
+            insertInto(CassandraMailboxRecentsTable.TABLE_NAME)
+                .value(CassandraMailboxRecentsTable.MAILBOX_ID, bindMarker(CassandraMailboxRecentsTable.MAILBOX_ID))
+                .value(CassandraMailboxRecentsTable.RECENT_MESSAGE_UID, bindMarker(CassandraMailboxRecentsTable.RECENT_MESSAGE_UID)));
+    }
+
+    public CompletableFuture<List<MessageUid>> getRecentMessageUidsInMailbox(CassandraId mailboxId) {
+        return cassandraAsyncExecutor.execute(bindWithMailbox(mailboxId, readStatement))
+            .thenApply(CassandraUtils::convertToStream)
```
Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentsDAO.java]: What is the added value of having only one stream function call wrapped in a thenApply call?

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentsDAO.java]: My idea is that it is more readable. 

And it allow more method reference, make it clearer we are working on the Future...

Comment (aduprat) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentsDAO.java]: I'm not sure it more readable, IMHO.

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 2]
```diff
@@ -89,23 +89,28 @@
...
-        return (mailbox -> ids.map(list -> list.contains(mailbox.getId())).orElse(true));
+    private Stream<Optional<Mailbox>> retrieveMailboxes(Optional<ImmutableList<MailboxId>> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+        if (mailboxIds.isPresent()) {
+            return mailboxIds.get().stream()
+                .map(mailboxId -> mailboxFactory.fromMailboxId(mailboxId, mailboxSession));
+        } else {
+            return retrieveUserMailboxes(mailboxSession)
+                .stream()
+                .map(MailboxMetaData::getPath)
+                .map(mailboxPath -> mailboxFactory.fromMailboxPath(mailboxPath, mailboxSession));
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: You could use only MailboxIds here too

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentDAOTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Test
+    public void removeFromRecentShouldOnlyRemoveUidWhenNotEmpty() throws Exception {
+        testee.addToRecent(CASSANDRA_ID, UID1).join();
+        testee.addToRecent(CASSANDRA_ID, UID2).join();
+
+        testee.removeFromRecent(CASSANDRA_ID, UID2).join();
+
+        assertThat(testee.getRecentMessageUidsInMailbox(CASSANDRA_ID).join()).containsOnly(UID1);
+    }
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentDAOTest.java]: can you add a test where you add the same uid twice?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java [hunk 4]
```diff
@@ -101,22 +102,17 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
         }
     }
 
-    private Stream<Optional<Mailbox>> retrieveMailboxes(Optional<ImmutableList<MailboxId>> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+    private Stream<MailboxId> retrieveMailboxes(Optional<ImmutableList<MailboxId>> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java]: please rename the method then

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentDAOTest.java [hunk 5]
```diff
@@ -96,4 +96,12 @@ public void removeFromRecentShouldOnlyRemoveUidWhenNotEmpty() throws Exception {
 
         assertThat(testee.getRecentMessageUidsInMailbox(CASSANDRA_ID).join()).containsOnly(UID1);
     }
+
+    @Test
+    public void addToRecentShouldBeIndepotant() throws Exception {
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxRecentDAOTest.java]: +m

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java [hunk 6]
```diff
@@ -104,7 +143,19 @@ private static boolean isChanged(Flags original, Flags updated, String userFlag)
     public Flags getOldFlags() {
         return oldFlags;
     }
-    
+
+    public boolean isSet(Flags.Flag flag) {
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: method name looks misleading, UpdatedFlags.isSet(recent) doesn't tell me that the recent flag changed from false to true.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java [hunk 7]
```diff
@@ -243,20 +247,31 @@ public void delete(MessageId messageId) {
     private CompletableFuture<Pair<MailboxId, UpdatedFlags>> updateCounts(Pair<MailboxId, UpdatedFlags> pair) {
         CassandraId cassandraId = (CassandraId) pair.getLeft();
         return CompletableFuture.allOf(
-            incrementCountIfNeeded(pair.getRight().getOldFlags(), pair.getRight().getNewFlags(), cassandraId),
-            decrementCountIfNeeded(pair.getRight().getOldFlags(), pair.getRight().getNewFlags(), cassandraId))
+            manageRecent(pair.getRight(), cassandraId),
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: updateRecent ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 8]
```diff
@@ -282,17 +297,33 @@ public MessageMetaData copy(Mailbox mailbox, MailboxMessage original) throws Mai
...
             return mailboxCounterDAO.incrementUnseen(mailboxId);
         }
-        if (!oldFlags.contains(Flag.SEEN) && newFlags.contains(Flag.SEEN)) {
+        if (updatedFlags.isSet(Flag.SEEN)) {
             return mailboxCounterDAO.decrementUnseen(mailboxId);
         }
         return CompletableFuture.completedFuture(null);
     }
 
+    private CompletableFuture<Void> manageRecents(Mailbox mailbox, UpdatedFlags updatedFlags) {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: looks like a pure logic duplication.
You probably miss a level of abstraction to be DRY

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMailboxCounterModule.java [hunk 9]
```diff
@@ -19,18 +19,19 @@
...
 import org.apache.james.backends.cassandra.components.CassandraTable;
 import org.apache.james.backends.cassandra.components.CassandraType;
 import org.apache.james.mailbox.cassandra.table.CassandraMailboxCountersTable;
 
-import java.util.Collections;
-import java.util.List;
-
-import static com.datastax.driver.core.DataType.counter;
-import static com.datastax.driver.core.DataType.timeuuid;
+import com.datastax.driver.core.schemabuilder.SchemaBuilder;
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMailboxCounterModule.java]: unrelated

**Comment by chibenwa** [comment]
O(n) to O(1) with JMAP

O(n) still with IMAP, but faster ^^

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java (hunk 6 above)
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: isSet during this flags update means this in my opinion.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java [hunk 10]
```diff
@@ -36,6 +37,44 @@
...
 
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private MessageUid uid;
+        private Flags oldFlags;
+        private Flags newFlags;
+        private long modSeq;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: Ok

File: mailbox/api/src/main/java/org/apache/james/mailbox/FlagsBuilder.java [hunk 11]
```diff
@@ -23,6 +23,10 @@
 
 public class FlagsBuilder {
 
+    public static FlagsBuilder builder() {
+        return new FlagsBuilder();
+    }
+
     private final Flags internalFlags;
 
     public FlagsBuilder() {
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/FlagsBuilder.java]: Not if I don't want to have a big code diff that will be hard to be merged. I would have normaly, but, let's put it this way, the team "have some pressure". Should I change it everywhere ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java [hunk 12]
```diff
@@ -71,22 +70,17 @@ public MailboxFactory(MailboxManager mailboxManager) {
...
         Optional<Role> role = Role.from(mailboxPath.getName());
-        MessageManager.MetaData mailboxMetaData = getMailboxMetaData(messageManager, mailboxSession);
         return Optional.ofNullable(Mailbox.builder()
                 .id(messageManager.getId())
                 .name(getName(mailboxPath, mailboxSession))
                 .parentId(getParentIdFromMailboxPath(mailboxPath, mailboxSession).orElse(null))
                 .role(role)
-                .unreadMessages(mailboxMetaData.getUnseenCount())
-                .totalMessages(mailboxMetaData.getMessageCount())
+                .unreadMessages(messageManager.getUnseenMessageCount(mailboxSession))
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java]: Because you have part of Metadata the list (arbitrary long) of all RECENT UIDs. This is the core of the problem (and it was not necessarilly a Cassandra design BTW)

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java (hunk 7 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageIdMapper.java]: If you want ;-)

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 8 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: As the one from SEEN count.

I will add to more mappers (and their tests) to care about that

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMailboxCounterModule.java (hunk 9 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/modules/CassandraMailboxCounterModule.java]: Ok. Due to @aduprat  automatic IMPORT reordering. Will separate in an other commit.

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java [hunk 13]
```diff
@@ -53,18 +53,24 @@
...
-     * @deprecated use
-     *             {@link #getMetaData(boolean, MailboxSession, org.apache.james.mailbox.MessageManager.MetaData.FetchGroup)}
      */
-    @Deprecated
     long getMessageCount(MailboxSession mailboxSession) throws MailboxException;
 
     /**
+     * Return the count of unseen messages in the mailbox
+     *
+     * @param mailboxSession
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java]: Ah OK, got it. Changed.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java (hunk 12 above)
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java]: And under the hood, you may do one call to the Mailbox layer, but still several, non parrallel calls where processed.

**Comment by chibenwa** [comment]
Rebased because some comments from @mbaechler implied reworking history

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java (hunk 12 above)
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MailboxFactory.java]: Ok, I will narrow my question : why don't we create a datastructure for Mailbox metadata that would not contain last uids ?

File: mailbox/api/src/main/java/org/apache/james/mailbox/FlagsBuilder.java (hunk 11 above)
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/FlagsBuilder.java]: If we don't have the time to do the change, the best choice is probably not doing it at all.
Do you really think anybody will handle the second half of the refactoring in some weeks ? IMO we will stay in a middle ground ...

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java (hunk 6 above)
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: it's not a flags update, it's an UpdatedFlags.
There's something wrong in the mix of present and past, I guess.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java [hunk 14]
```diff
@@ -95,6 +98,7 @@ private PreparedStatement updateMailboxStatement(Session session, Assignment ope
     }
 
     public CompletableFuture<Void> decrementUnseen(CassandraId mailboxId) {
+        new Exception().printStackTrace();
```
Comment (rouazana) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAO.java]: Hmmm

File: mailbox/api/src/main/java/org/apache/james/mailbox/FlagsBuilder.java (hunk 11 above)
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/FlagsBuilder.java]: I took time

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java (hunk 6 above)
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: Moved to past

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java [hunk 15]
```diff
@@ -146,16 +146,16 @@ public Flags getOldFlags() {
         return oldFlags;
     }
 
-    public boolean isSet(Flags.Flag flag) {
+    public boolean wasSet(Flags.Flag flag) {
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: that's worst for me: wasSet should mean that only oldFlag is concerned, and that oldFlag contains it.
maybe isModifiedToSet? (don't find better, sorry)
and this other should be then isModifiedToUnset and isUnchanged.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java [hunk 16]
```diff
@@ -36,6 +38,45 @@
...
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private MessageUid uid;
+        private Flags oldFlags;
+        private Flags newFlags;
+        private Optional<Long> modSeq = Optional.absent();
+
```
Comment (strokyl) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: Aren't supposed to make the constructor private since there is a static method above to construct the builder

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java (hunk 6 above)
Comment (strokyl) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: I agree with @mbaechler maybe hasJustBeenSet would be a better name. But I am bad at naming stuff that speak to other people than myself.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java [hunk 17]
```diff
@@ -104,7 +145,19 @@ private static boolean isChanged(Flags original, Flags updated, String userFlag)
     public Flags getOldFlags() {
         return oldFlags;
     }
-    
+
+    public boolean isSet(Flags.Flag flag) {
+        return newFlags.contains(flag) && !oldFlags.contains(flag);
+    }
+
+    public boolean isUnset(Flags.Flag flag) {
```
Comment (strokyl) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: Same as above

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java (hunk 6 above)
Comment (strokyl) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: Or maybe naming the class DiffFlag would be better than UpdatedFlag

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java (hunk 17 above)
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: Subtle difference here : negation is not at the same place ...

```
    public boolean isModifiedToSet(Flags.Flag flag) {
        return newFlags.contains(flag) && !oldFlags.contains(flag);
    }

    public boolean isModifiedToUnset(Flags.Flag flag) {
        return !newFlags.contains(flag) && oldFlags.contains(flag);
    }
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java (hunk 16 above)
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/UpdatedFlags.java]: I concider it is boiler plate and do not add much value IMO.

Making the constructor of the built class private is a hudge win as you force using the builder, with explicit naming, and force builder adoption. Here, the choice between builder constructor and static method is just about fluency (they are strictly equivalent).

Anyway, while squashing, I will make it private.

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAOTest.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+    @Test
+    public void countUnseenMessagesInMailboxShouldReturnEmptyByDefault() throws Exception {
+        assertThat(testee.countUnseenMessagesInMailbox(mailbox).join().isPresent()).isFalse();
+    }
+
+    @Test
+    public void incrementCountShouldAddOneWhenAbsent() throws Exception {
+        testee.incrementCount(MAILBOX_ID).join();
+
+        assertThat(testee.countMessagesInMailbox(mailbox).join().get()).isEqualTo(1);
```
Comment (chibenwa) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/CassandraMailboxCounterDAOTest.java]: Forgot. Anyway, it was just to make exception more readable. It will have no effect on test. Just error instead of failure.
